controllers/clusterurlmonitor/clusterurlmonitor.go

```diff
@@ -0,0 +1,123 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clusterurlmonitor
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/openshift/route-monitor-operator/api/v1alpha1"
+	monitoringv1alpha1 "github.com/openshift/route-monitor-operator/api/v1alpha1"
+	"github.com/openshift/route-monitor-operator/pkg/blackboxexporter"
+	utilreconcile "github.com/openshift/route-monitor-operator/pkg/util/reconcile"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+)
+
+// ClusterUrlMonitorReconciler reconciles a ClusterUrlMonitor object
+type ClusterUrlMonitorReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+const (
+	FinalizerKey string = "clusterurlmonitor.monitoring.openshift.io/clusterurlmonitorcontroller"
+)
+
+// +kubebuilder:rbac:groups=monitoring.openshift.io,resources=clusterurlmonitors,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=monitoring.openshift.io,resources=clusterurlmonitors/status,verbs=get;update;patch
+
+func (r *ClusterUrlMonitorReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	log := r.Log.WithValues("clusterurlmonitor", req.NamespacedName)
+
+	// your logic here
+	clusterUrlMonitor, res, err := r.GetClusterUrlMonitor(ctx, req)
+	if err != nil {
+		return utilreconcile.RequeueWith(err)
+	}
+
+	if res.ShouldStop() {
+		return utilreconcile.Stop()
+	}
+
+	blackboxExporter := blackboxexporter.NewBlackboxExporter(r.Client, log, ctx)
+	s := NewSupplement(clusterUrlMonitor, r.Client, r.Log, blackboxExporter)
+
+	res, err = s.EnsureFinalizer()
+	if err != nil {
+		return utilreconcile.RequeueWith(err)
+	}
+	if res.ShouldStop() {
+		return utilreconcile.Stop()
+	}
+
+	err = blackboxExporter.EnsureBlackBoxExporterResourcesExist()
+	if err != nil {
+		return utilreconcile.RequeueWith(err)
+	}
+
+	serviceMonitor, err := s.EnsureServiceMonitor()
+	if err != nil {
+		return utilreconcile.RequeueWith(err)
+	}
+
+	res, err = s.EnsureDeletionProcessed(serviceMonitor)
+	if err != nil {
+		return utilreconcile.RequeueWith(err)
+	}
+	if res.ShouldStop() {
+		return utilreconcile.Stop()
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ClusterUrlMonitorReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&monitoringv1alpha1.ClusterUrlMonitor{}).
+		Complete(r)
+}
+
+// GetClusterUrlMonitor return the ClusterUrlMonitor that is tested
+func (r *ClusterUrlMonitorReconciler) GetClusterUrlMonitor(ctx context.Context, req ctrl.Request) (v1alpha1.ClusterUrlMonitor, utilreconcile.Result, error) {
```

Comment:
should be seperated to a supplement file?

Reply:
yes